### ISSUE TITLE
terapadog

### ISSUE THREAD

**Issue by Gionmattia** [Bioconductor/Contributions#3615]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/Gionmattia/terapadog

Confirm the following by editing each check box to '[x]'

- [ x ] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [ x ] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [ x ] I understand Bioconductor [Package Naming Policy][9] and acknowledge
  Bioconductor may retain use of package name.

- [ x ] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [ x ] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [ x ] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.
  
- [ x ] I am familiar with the [Bioconductor code of conduct][7] and 
  agree to abide by it.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [ x ] The 'devel' branch for new packages and features.
- [ x ] The stable 'release' branch, made available every six
      months, for bug fixes.
- [ x ] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For questions/help about the submission process, including questions about
the output of the automatic reports generated by the SPB (Single Package
Builder), please use the #package-submission channel of our Community Slack.
Follow the link on the home page of the [Bioconductor website][8] to sign up.

[1]: https://contributions.bioconductor.org/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/
[7]: https://bioconductor.org/about/code-of-conduct/
[8]: https://bioconductor.org/
[9]: https://bioconductor.org/developers/package-submission/#naming

**Comment by lshep** [comment]
We recommend changing the name of vignette.Rmd to something more informative as
there can be naming conflicts depending on package load order.

id_converter should not automatically write to the home directory.  You should
have an argument option on where this file should be written that should default
to a tempdir(). Similar for plotDTA

You mention how its integration with ReactomeGSA is where it truly shines but I
don't see that demonstrated in the vignette?  currently I don't see the interoperability?

**Comment by Gionmattia** [comment]
Hi,

Thank you for the feedback! Here’s my response, point by point:

I renamed vignette.Rmd to something more straightforward (but not particularly creative): terapadog_vignette.Rmd.

For id_converter, I followed your advice and added an argument that defaults to a tempdir().
Regarding plotDTA, the issue was with the runnable example: during the checks, it was automatically generating a plot in the home directory (my oversight).
I resolved this by correcting the example and ensuring the plot is saved to a tempdir() by default if no path is provided.

About the ReactomeGSA integration: I may not have explained this part clearly earlier, so let me start from the beginning:

Terapadog is designed as a package with functions integrated into the ReactomeGSA backend (https://github.com/reactome/ReactomeGSA/tree/87-ribo-seq). Its purpose is to perform Gene Set Enrichment Analysis using Ribo-Seq data on the Reactome web platform.

Currently, it is integrated within the development branch (87-ribo-seq) of the ReactomeGSA repository and will be pushed to main once Terapadog is approved by Bioconductor.

Since I was asked to release Terapadog as a Bioconductor package, I decided to include additional functions to allow it to perform the "core" of the analysis independently. This means it can analyze data locally if users are interested in identifying enriched pathways undergoing translational regulation. However, its full potential is realized when used within the Reactome portal, where it can take advantage of interactive pathway visualizations and other tools.

I didn’t include an example of this integration in the vignette because:

The implementation occurs outside the R environment.
I planned to include a link (and relevant Reactome tutorial page) once the package and integration are officially released.
I believe the main point of confusion might be the unilateral nature of the integration: while the ReactomeGSA backend loads functions from Terapadog, the reverse is not true—Terapadog does not load ReactomeGSA functions.

I’ve updated the explanation in terapadog_vignette.Rmd to make this section clearer. I hope this addresses any confusion.

Thank you again for your helpful feedback!

**Comment by lshep** [comment]
I believe you have misspelled rmarkdown dependency in your vignette.  'rmakrkdown'

**Comment by js2264** [comment]
Hi @Gionmattia , sorry for the long wait, I was away for some time during the winter break. Here is my review of your package: 

---

The package provides a set of functions to perform differential translation analysis. Documention is extensive, and the vignette provides a good overview of the package and its functionality. 

However, the main `terapagod()` function is never actually executed, neither in the vignette, tests, or examples. It would be good to have some tests for this function, as it is the main function of the package.

I could not actually execute it myself for now, as Ensembl service seems to be unavailable as of today. I'll try again later. For now, here are my comments: 


## General 

- [ ] `LazyData` field in `DESCRIPTION`

This should be set to `FALSE`, unless you have a good reason to set it to `TRUE`? If so, please provide a justification.

- [ ] `BugReports` and `URL` fields in `DESCRIPTION`

Please provide `URL` and `BugReports` fields. `URL` can point to either the GitHub repository or the Bioconductor package landing page. `BugReports` should point to the issue tracker of the package (most likely GitHub).

- [ ] `NEWS.md` file

The `NEWS.md` file in the `inst/` directory does not look like it's kept up to date. Please update it with the latest changes to the package. If you are not planning to keep it up to date, please remove it.

## Vignette 

- [ ] Why is `manual_example.html` present in the `vignettes/` directory? I understand it is being generated when building the vignette, but it should not be present in the `vignettes/` directory. Please remove it from the `vignettes/` directory, and add it to the `.gitignore` file.

- [ ] `fc_results <- get_FCs(expression.data, exp_de)` (line 415) in the vignette fails if `apeglm` is not installed. I would recommend adding a `require()` call to `apeglm` prior to this command. 

## Tests 

- [ ] Please consider at least a few tests for your package (see [here](https://contributions.bioconductor.org/tests.html) for more information). You can use the `testthat` package to write tests. I strongly advise you to write tests for your package, as it will help you catch bugs early and ensure that your package works as expected. An easy start would be to wrap some of your examples in unit tests, for example. 

## R code

- [ ] `id_converter()`: `biomaRt` requires the Ensembl server to be up and running (which seems to not be the case right now!). Could you perhaps rely on an Annotation package, rather than biomaRt (this would require to add `AnnotationDbi` and `org.Hs.eg.db` packages as `Suggests` or `Imports`)? Something like (untested): 

```r
AnnotationDbi::mapIds(org.Hs.eg.db::org.Hs.eg.db, ...)
```

- [ ] `plotDTA.R` line 25: the `path = paste0(tempdir(), "/plot.html")` argument will fail in platforms where the path separator is not `/` (windows). I would recommend using `file.path(tempdir(), "plot.html")` instead.

- [ ] `terapadog.R` line 90 and others: whenever you want to print an informative message to the R console (`if (verbose) {...}`), please use dedicated functions, eg. `message()` or `warning()`. Do not use `paste0()/paste()` in these methods, as it is not needed. 

- [ ] `check_value_range()`: perhaps I am missing smthg here, but wouldn't it be safer to assess whether the provided matrix is numerical?

**Comment by Gionmattia** [comment]
Good morning Jacques,

Thanks for the feedback. It took me more time than expected to run through
all the corrections (sorry for the wait).
Below is a summary of the changes:


*General:- *Set LazyData field to FALSE
- Added URL and BugReports as specified, within the DESCRIPTION file
- Removed NEWS.md file (not really useful at this stage)

*Vignette:*
- Fixed the issue with plotDTA and the manual_example.html file generated
by the vignette. saved. The function now saves the plot only if explicitly
requested.
- Added a check to see whether apeglm is installed or not, before running
related commands.


*Test:- *Added unit tests (tests/ folder).

   - tests cover the various functions in the package, ensuring the
   input/outputs (or critical steps) are carried without major issues.
   - Let me know if you think there are other aspects you think should be
   tested.

*R code:*
- Implemented a local conversion table from Ensembl, with its version
specified in inst/script/extdata_info.md document.

   - id_converter() now first checks the status of Ensembl (tryCatch).
   - If Ensembl is down, a warning message is issued, and the local table
   is used.
   - given that id_converter() is mainly intended for standalone use, I
   believe this should suffice. The whole package is conceived to be
   integrated by the ReactomeGSA backend, which already handles ID conversions
   internally. This is a way around that will allow users to still be able to
   run if Ensembl is down, and the conversion table will be updated with the
   latest annotation releases.
   - However, if user feedback suggests this remains a common issue (the
   unreliability of Ensembl servers) I will consider the switch to another
   package instead.

- Fixed the issue with file.path()
- Updated terapadog to use message() instead of paste0().
- Regarding check_value_range(): This function ensures the
user-provided data is NOT normalised between 0 and 1.

   - It is specified in the documentation that raw count data is required,
   but if a user proceeded anyway, the analysis would run anyway, but with
   unreliable results.
   - Data between 0 and 1 is an edge-case this function is accounting for.

Let me know if any further refinements are needed: It is my first time
publishing a package in bioconductor, so every step of the process is new
to me and I appreciate the feedback!

Best regards,

Gionmattia

**Comment by js2264** [comment]
Hi @Gionmattia , thank you for your changes. I have been through the updated package version and below are my comments. I see that you have already implemented many tests, which is great, but I would still need to see the `terapadog()` function being tested, to be able to add it to Bioconductor, unless you tell me that this function is not central to the aim of the package. Let me know if you need any help in implementing tests, I'll be happy to provide guidance. 

---

- One thing I don't really undersand (and I cidn't notice earlier, sorry) is that in the vignette, you call `results <- terapadog(...)`, but the rest of the vignette focuses on `fc_results` object, which does not seem to be the output of `terapadog()`. So what is the use of the `results` object? and compared to `fc_results`? How is the end-user meant to use it? I think at this point it would be really useful to actually evaluate the call to `terapadog()` and show the output. Is there a good reason for not doing so? 

>  - tests cover the various functions in the package, ensuring the input/outputs (or critical steps) are carried without major issues. 
> - Let me know if you think there are other aspects you think should be tested.

The `terapadog()` function is still not tested, even though it seems to be the cornerstone of the package. It's also not evaluated in the vignette, so a bug is introduced, you will have no way to know it. This often occurs even without updates to YOUR code base, but because upstream dependencies are updated themselves and these changes introduce bugs in your own function. If you really do not wish to evaluate all the code in the vignette, I would need at least the `terapadog()` function to be tested during tests, to make sure that it is future-proof. 

> - Added a check to see whether apeglm is installed or not, before running related commands.

I see that you have explicitely added this check in the vignette. This will help the end-user who reads your vignette. Perhaps a less verbose/more canonical way would have been simply to run `require(apeglm)`, but your approach works as well. 

> - Implemented a local conversion table from Ensembl, with its version specified in inst/script/extdata_info.md document.

This works for now, but I don't really understand why you wouldn't want to rely on using the dedicated `org.Hs.eg.db` package. Its primary goal is to provide an updated gene ID database for human genes, and support conversion between different types of IDs. But I do realize that its syntax might not be very easy to understand at first. I would recommend you read about this for future projects or releases of this package, I'm sure you'll find that it can do many things you wanted to do.

**Comment by Gionmattia** [comment]
Hi Jacques,

Thanks for the feedback. I wanted to quickly clarify a few points, before
getting back at improving my package.


*About the results:*
The terapadog() function applies the PADOG algorithm to the results of
a *Differential
Translation Analysis *(DTA), specifically performing a *pathway enrichment
analysis *on genes identified as undergoing translational regulation due to
changes in their *translational efficiency *(TE). This workflow has two key
outputs:

   - *Pathway-level results:* obtained via terapadog(), showing the
   enriched pathways and related genes.
   - *Gene-level results (and plot): *obtained through the get_fc()
   function, which displays each gene and their regulatory mode (RegMode).

So these are two different ways to look at the analysis results. From a
perspective of integration within the ReactomeGSA back-end, terapadog() is
more important, while get_fc() is to further contextualise the results for
a human user.

I do think there could be space for further integrations between the two
results (maybe in the form of interactive visualisations), but due to time
constraints I found myself having to postpone these to a future update (the
struggle of a PhD: too many things to do, too little time).

*About terapadog() running in the vignette:*

The main reason I avoided executing terapadog() within the vignette is that
it performs permutation tests, so I am afraid of its computational cost
(and that it would slow down the installation of the package).I aimed at
ensuring a quick installation, while still providing users with datasets
and tutorials so they can run the analysis themselves.

That said, I see that there's a need for further clarification of the
results within the vignette. Would it be acceptable if I expanded the
explanation of terapadog()'s results, without executing it?


*About unit testing for terapadog() itself:*
I see your point, I focused mostly on the other functions to make sure
inputs/outputs were clearly formatted, but what you say is true - I need to
ensure terapadog() itself is "robust". Again, this is my first time writing
a package, so thanks for taking time to explain these bits to me. It's easy
to overlook something, when inexperienced.
I would rather run terapadog() for tests then, rather than when building
the vignette.

*Next steps:*
I will add unit tests for terapadog() itself, so to make sure it runs (and
some of its core functionalities run correctly as well).
I will improve the vignette documentation, to explain the relationship
between the results.

Let me know if this approach is ok or if I am missing something.

Best Regards,

Gionmattia

**Comment by Gionmattia** [comment]
Good morning @js2264,

I have updated the package to address your latest comments. Briefly:

- I have expanded the vignette to provide more detailed explanations about the different outputs (the simple DTA and the GSEA) and included examples. Since running terapadog() is computationally intensive, I have added pre-computed results to the package, which can be loaded within the vignette as an example. This should help users better understand the output, and the breakdown further clarifies what each metric represents.

- I have added tests for terapadog(). Given that this analysis is **computationally intensive**, I had to optimize the tests to run within Bioconductor's server constraints, avoiding timeout errors. To achieve this, I first restructured the entire terapadog() function, which now calls three subfunctions (found in "terapadogBricks.R"), each with its own internal checks and tests. Additionally, I designed tests that operate on smaller subsets of data with minimal parameters. This rework required a considerable amount of effort.

I hope this update will suffice in addressing your previous concerns.

Best,
Gionmattia

**Comment by js2264** [comment]
Hi @Gionmattia , thank for your reply, I'll have a look tomorrow, but the bullet list you wrote sounds perfect to me. I'm certain everything will be ok. I'll ping you tomorrow after review. Thank you for your efforts!

**Comment by js2264** [comment]
Hi @Gionmattia , I have been through the changes you have made. I greatly appreciate the improvements, I believe that they significantly strengthen the package infrastructure from a developer standpoint. Breaking `terapadog()`  function in smaller bits is a great approach to make your code more robust and testable. FYI: you went from 38% test coverage in 0.99.4 to 78% in 0.99.6!!

```{r}
> covr::package_coverage('.')
terapadog Coverage: 78.63%
R/get_FCs.R: 27.03%
R/id_converter.R: 58.14%
R/terapadogBricks.R: 84.46%
R/preprocessing_helpers.R: 88.00%
R/terapadog.R: 93.33%
R/prepareTerapadogData.R: 97.96%
R/assign_Regmode.R: 100.00%
R/plotDTA.R: 100.00%
```

The changes in the vignette also clarify a lot the purpose of the package, to me, thank you for these changes. 

I am happy with the state of the package, thank you for your contribution to Bioconductor!

**Comment by lshep** [comment]
The default branch of your GitHub repository has been added to Bioconductor's
git repository as branch devel.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/Gionmattia.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("terapadog")`. The package 'landing page' will be created at

https://bioconductor.org/packages/terapadog

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
